# CITE-SEQ Analysis in MSI-H Classified Breast Cancer Patients

This is an implemntation of CITE-Seq analysis as it relates to msi cancer detection.The primary use case of this pipeline is to coorlate cell surface markers with MSI-H classified patients and determine potential biomarkers of therapeutics relevance.

To assess the feasibility of this study and hypothesis we explored a synergistic approach of scRNA-Seq data with CITE-Seq (Transcriptomic and Proteomic). To that end, we used publicly available data from The Breast Cancer Atlas<sup>[24](https://www.nature.com/articles/s41588-021-00911-1)</sup>. This dataset is a wealth of resources for single cell expression data in human breast cancer. Consisting of 100K cells, it represents a variety of different cell states and diversity within tumor microenvironment. We utilize `msi-sensor`, a SVM model that detects replication slippage variants at microsatellite regions in both bulk and single-cell expression data, to classify individual cells as MSS or MSI-H <sup>[46](https://github.com/xjtu-omics/msisensor-rna)</sup>. We use `scanpy`, a single-cell expression analysis package, to analyze the 10X data incorporating quality control, normalization and clustering <sup>[47](https://github.com/scverse/scanpy)</sup>. Finally, we use the normalized expression matrix to form the foundational train – test dataset for our XGBoost model <sup>[48](https://learning.oreilly.com/library/view/hands-on-machine-learning/9781492032632/)</sup>.

There are some major caveats and limitations to this pilot data:

1. MSI-Sensor is pre-trained on MSS and MSS-H datasets<sup>[46](https://academic.oup.com/gpb/article/22/3/qzae004/7515254)</sup>. While selected for its toleration of single cell data, there is not a pre-trained model on breast cancers and the model selected, `TCGA.MSIPopular.model.pkl`, was trained on CRC (colorectal cancer), STAD (stomach adenocarcinoma), and UCEC (uterine corpus endometrial carcinoma). While there may be some similarities, these are different cancers with unique transcriptomic landscapes and may not be the most representative model.

2. This dataset was initially selected for its inclusion of CITE-Seq data, which at first was assumed to be multi-modal with the scRNA data, but upon data exploration found that it was in fact entirely separate<sup>[24]()</sup>. This creates a barrier for direct correlation of surface protein stains and gene expression profiles in the same cell. The subsequent CITE-Seq analysis should be treated as a bulk patient population snapshot rather than a direct one-to-one comparison.

3. The publicly available metadata associated with this patient population is extremely limited. We cannot be certain what phase these patients are, if they have responded positively to any immune checkpoint inhibitors, or if they have been classified as MSS-H or MSS. Therefore, it's difficult to say with confidence whether or not our analysis aligns with their medical diagnosis. We can merely report total probability of MSI based on sub-clonal population probability density.

As a data exploration task, we aimed to predict cell types based on specific marker genes associated with stratified cell types (e.g CD4 with CD4+ T-cells). While MSI can impact any cell, MSI-H within the TME shows a particular enrichment to CD8+ T-Cells, M1 Macrophages and myeloid-derived suppressor cells (MDSCs). Inflamed TME of MSI-H tumors correlates with higher mutational burden and immune infiltration and inflammation. Several subtypes including B-cells, stromal and myeloid cells were uniquely and distinctly clustered using UMAP. Other types including T-Cells and the cancer subtypes, had more overlap indicating confounding feature scores between those cell types, based on their gene scoring panels. Notably, we observed that cancer subtypes (basal and HER2+) displayed significant transcriptional diversity, resulting in minimal clustering when projected alongside immune populations. This phenomenon could be explained by dynamic cell states in cancer cells, potential epithelial-mesenchymal transition events occurring within the tumor ecosystem, or an incomplete gene panel.

---

A brings up a particularly intriguing prospect for this proposal - the dramatically heterogenous nature of MSI predictions within an individual patient’s tumor cells. The tumor microenvironment is exceptionally dynamic and consists of a pantheon of subtypes. While a patient may be diagnosed with MSS/MSI-H, their sub-clonal population may be more nuanced, giving rise to a patient population that might respond well to PD-1 inhibitors that would have been previously overlooked. This prediction is the crux of this data. Being able to correlate MSI status with cell-surface markers would enable specific solid tumor targeting with CAR-T delivery of PD-1 inhibitors, leading to lower doses, higher on target efficacy and lower toxicity of the PD-1 payload.

To estimate the underlying `msi_status` in data-exploration, several genes were selected as a scoring panel: WRN, MLH1, MLH2, MLH3, genes involved in DNA metabolism and repair. This matrix was fairly uninformative. No distinct difference in `msi_status` could be detected and they were scrapped in favor of a gene panel isolated from DEGs based on `msi_status` itself. Several overly expressed DEG in MSS-H were selected and identified as being 2 - 8-fold enriched in MSS-H cells: DUSP4, TNFAIP3, TNFRSF18, TNFSF9 and GLNY.

We used the above 5 genes to score the MSI status as a correlative measure against the msi-sensor pipeline. In truth, because the `msi_status` was predicted from this pipeline in the first place, these 5 genes represent the highest importance gene features for the MSI status classification, and the simple scanpy scoring assignment represents something of a distillation of that SVM model. Nevertheless, we were curious that these markers could truly correlate highly with the 105 gene model alone on simple average expression set of these genes. What we found was a strong correlation between the probability of MSI-H and the msi_score represented by these genes, meaning they can serve as marker genes in subsequent analysis.

By classifying these patients as high risk we can turn to the CITE-Seq data and analyze their surface protein expression identifying any surface expression patterns. This is where we feel directed funding could unlock cell state predictions. To reiterate, because the `msi_status` of patients is unknown in the breast cancer atlas, all we can do is speculate on their status and use a biologically similar model. This is not ideal. Biopsies from patients diagnosed with breast cancer and MSS-H would be used in the 10X Immune characterization kit. This would link single cell transcription with single cell surface patterns. Once the cell surface patterns are known, biopsies from patients could be merely passed to either an ELISA or AB stain for clinicians to make a quantitative diagnosis.

One important linkage is the immune/inflammation modulations. We hypothesize based on publicly available research that we could score cells for these immune markers, and they would correlate with the msi_score. For our first pass with a logistic regression model, we use the aggregate scores that we generated from our gene panel in scanpy : msi_score, immune_score, and stromal_score. The main purpose of this approach is to determine the predictive potential of our scores against the SVM output. What we found is that both as the immune score increases, so too does the msi_score. Similarly, the msi_score follows a trajectory like the SVM msi_probability. We see a Spearman correlation of 0.61 indicating the smaller gene panel scoring reliably with the 105 SVM gene panel. Unfortunately, this model has terrible specificity (55%). Clearly, a simple scoring matrix is insufficient to capture the cellular heterogeneity of a tumor sample. Some research suggests that panels as large as 0.2 - 0.6 Mb are needed to show accurate TMB/MSI prediction.

We believe this model’s poor performance came from several fundamental constraints:
1.	Its linearity assumption was wrong and failed to capture the complex, non-linear relationships between gene expression patterns in heterogeneous tumor micro-environments.
2.	Its reliance on aggregated gene panel scores (msi_score, immune_score, stromal_score) proved insufficient for distinguishing true MSI signals from confounding myeloid and stromal signatures.
3.	Feature contributions were unweighted, leading to high false positives, particularly in samples with elevated immune infiltration independent of MSI status.

Ultimately, this was scrapped in favor of a model that might perform better - 55% false negatives is nowhere near sufficient to inform a patient. We opted to train a XGBoost model on the expression matrix using the `msi_status` as the prediction outcome. Essentially, this is a distillation of the SVM model that was used in msi-sensor, but it is still a good exercise to rank the feature importance and potentially identify any cell surface markers that are expressed highly in the CITE-Seq data. XGBoost was chosen for its ability to both classify and model non-linear relationships. From the trained model we can potentially identify genes which are most predictive of MSI status.

![alt text](https://github.com/nwlodychak/cite_seq/blob/main/results/msi-signature-heatmap.png?raw=true "DEG by MSI signature")

Unlike our simple and limited logistic regression model, XGB captures the complex interdependencies between expression patterns that might be missed in more simplistic models, including a resistance to cellular heterogeneity that may be in the tumor microenvironment. We found this model to independently identify many biologically significant makers without prior constraint. We identified several markers that were differentially expressed in MSS-H (predicted) and MSS (predicted) cell states. DUSP2 and DUSP4 (ρ = 0.72) are well known expression markers for MSS-H cancers. Both parts of the MAPK pathway, they are linked to highly metastic cancers and especially MSI-H colorectal cancer. Other gene markers of interest are TNFAIP3 and CST7 which are both immune/inflammation modulators. The former regulates NF-kB signaling while the latter inhibits proteases in cytotoxic lymphocytes leading to potential immune evasion. These markers correlate with the hypothesis that immune infiltration makers correlate with the msi_score. Additionally, we observed substantially better performance - 95.4% accuracy and 95.5% specificity— compared to initial logistic regression approach that yielded only 55% specificity.


![alt text](https://github.com/nwlodychak/cite_seq/blob/main/results/predicted-outcomes.png?raw=true "XGBoost Confusion Matrix")

![alt text](https://github.com/nwlodychak/cite_seq/blob/main/results/msi-deg.png?raw=true "MSI Differentially Expressed Genes (DEG)")

Several exciting approaches with CAR-T have shown promise in patients with solid tumors. Genes from our isolated gene panel are already being developed and paired with relevant cancer therapies. TNFSF9 for example, binds to 4-1BB on CAR-T cells and enhances their metabolic fitness reducing exhaustion and improving their persistence in solid tumors through co-stimulatory pathways <sup>[49](https://pmc.ncbi.nlm.nih.gov/articles/PMC8773791/),[50](https://pubmed.ncbi.nlm.nih.gov/34706886/)</sup>. TNFAIP3, in a tm-TNF-⍺ CAR-T study found that TNFAIP3 medicated NF-kB inhibition up-regulated PD-L1 <sup>[49](https://pmc.ncbi.nlm.nih.gov/articles/PMC8773791/),[51](https://www.pnas.org/doi/10.1073/pnas.1406259111)</sup>. We believe that this could be a missing piece from our logistic regression model, limiting its specificity due to PD-L1+ stromal interference. Combining anti-PD-1 antibodies with CAR-Ts improved tumor regression 68% in some patients <sup>[52](https://pubmed.ncbi.nlm.nih.gov/35919489/)</sup>. Finally, DUSP4 activity is consistent with studies where ERK inhibition reduced CAR-T granzyme B release by up to 60% <sup>[53](https://doi.org/10.1101/2024.03.18.585442)</sup>. Some research suggests that co-expression of DUSP4 and HER2 in MSI-H sub-clones means that HER2 directed CAR-Ts could bypass this resistance as validated in HER2+ breast cancer, a key point of the underlying dataset <sup>[54](https://pubmed.ncbi.nlm.nih.gov/36831412/),[55](https://pmc.ncbi.nlm.nih.gov/articles/PMC5652774/)</sup>.

![alt text](https://github.com/nwlodychak/cite_seq/blob/main/results/dusp4-signature.png?raw=true "DUSP4 Signature Score")

MSI status prediction based on minimal data input is lucrative for several reasons. Patients with MSI-H presentation better respond to immune checkpoint inhibitors such as PD-1/PD-L1 blockade and respond worse the typical chemotherapy drugs <sup>[23](https://pmc.ncbi.nlm.nih.gov/articles/PMC10700246/), [56](https://pubmed.ncbi.nlm.nih.gov/37686520/)</sup>. The fact that many immune evasion markers were detected significantly in this dataset supports the hypothesis of developing dual-targeting approaches and precision immunotherapies. Possible outcomes and developmental strategies following this dataset exploration pilot include dual-targeting CAR-Ts, co-targeting HER2 and TNFSF9 could exploit the MSI-H vulnerabilities while simultaneously utilizing 4-1BB co-stimulation <sup>[7](https://pubmed.ncbi.nlm.nih.gov/37894457/), [20](https://pubmed.ncbi.nlm.nih.gov/36969740/), [50](https://pubmed.ncbi.nlm.nih.gov/34706886/), [57](https://pubmed.ncbi.nlm.nih.gov/39938523/)</sup>. Additionally, these therapies in combination with Anti-PD-1 antibodies may reverse the TNFAIP3-causal immunosuppression that has been recorded in some tmTNF-⍺ CAR-T trials <sup>[18](https://pubmed.ncbi.nlm.nih.gov/36720496/)</sup>.
The approach presented only somewhat scratches the surface of translation with this dataset. With further evaluation, finances, and time we might utilize tools like sciPENN, a deep learning approach, for combining diverse CITE-Seq and scRNA-Seq batches, to elucidate more informative panels <sup>[19](https://pmc.ncbi.nlm.nih.gov/articles/PMC9979929/)</sup>. Highly desirable information garnered from supplemental studies on these patients, especially from temporal trajectory of diagnosis could better inform patient dosing schedules.

## References

1.	van Wietmarschen, N. et al. Repeat expansions confer WRN dependence in microsatellite-unstable cancers. Nature 586, 292–298 (2020).
2.	Xie, H. & Zhang, J. Targeting werner helicase with synthetic lethality in microsatellite instability cancers: Promising therapeutic approaches. MedComm – Oncol. 3, e82 (2024).
3.	Nádorvári, M. L., Lotz, G., Kulka, J., Kiss, A. & Tímár, J. Microsatellite instability and mismatch repair protein deficiency: equal predictive markers? Pathol. Oncol. Res. 30, 1611719 (2024).
4.	Mulet-Margalef, N. et al. Challenges and Therapeutic Opportunities in the dMMR/MSI-H Colorectal Cancer Landscape. Cancers 15, 1022 (2023).
5.	MMRd, MSI-H, and TMB-H Tumors: What They Are and Why They Matter for Cancer Immunotherapy | Memorial Sloan Kettering Cancer Center. https://www.mskcc.org/cancer-care/diagnosis-treatment/cancer-treatments/immunotherapy/mmrd-msi-h-and-tmb-h-tumors.
6.	Zong, D. et al. Comprehensive mapping of cell fates in microsatellite unstable cancer cells supports dual targeting of WRN and ATR. Genes Dev. 37, 913–928 (2023).
7.	Heregger, R. et al. Unraveling Resistance to Immunotherapy in MSI-High Colorectal Cancer. Cancers 15, 5090 (2023).
8.	Sobol, R. W. WRN suppresses p53/PUMA-induced apoptosis in colorectal cancer with microsatellite instability/mismatch repair deficiency. Proc. Natl. Acad. Sci. 120, e2219963120 (2023).
9.	Chan, E. M. et al. WRN helicase is a synthetic lethal target in microsatellite unstable cancers. Nature 568, 551–556 (2019).
10.	Sommers, J. A. et al. A high-throughput screen to identify novel small molecule inhibitors of the Werner Syndrome Helicase-Nuclease (WRN). PloS One 14, e0210525 (2019).
11.	Ferretti, S. et al. Discovery of WRN inhibitor HRO761 with synthetic lethality in MSI cancers. Nature 629, 443–449 (2024).
12.	Branca, M. WRN May Be Key to Overcoming Colorectal Cancer Drug Resistance. Inside Precision Medicine https://www.insideprecisionmedicine.com/news-and-features/wrn-may-be-key-to-overcoming-colorectal-cancer-drug-resistance/ (2021).
13.	Morales-Juarez, D. A. & Jackson, S. P. Clinical prospects of WRN inhibition as a treatment for MSI tumours. Npj Precis. Oncol. 6, 1–6 (2022).
14.	Dedeurwaerdere, F. et al. Comparison of microsatellite instability detection by immunohistochemistry and molecular techniques in colorectal and endometrial cancer. Sci. Rep. 11, 12880 (2021).
15.	Kamath-Loeb, A. S., Loeb, L. A., Johansson, E., Burgers, P. M. & Fry, M. Interactions between the Werner syndrome helicase and DNA polymerase delta specifically facilitate copying of tetraplex and hairpin structures of the d(CGG)n trinucleotide repeat sequence. J. Biol. Chem. 276, 16439–16446 (2001).
16.	Pirzio, L. M., Pichierri, P., Bignami, M. & Franchitto, A. Werner syndrome helicase activity is essential in maintaining fragile site stability. J. Cell Biol. 180, 305–314 (2008).
17.	Andreu-Saumell, I. et al. CAR affinity modulates the sensitivity of CAR-T cells to PD-1/PD-L1-mediated inhibition. Nat. Commun. 15, 3552 (2024).
18.	Ba, H. et al. Antitumor effect of CAR-T cells targeting transmembrane tumor necrosis factor alpha combined with PD-1 mAb on breast cancers. J. Immunother. Cancer 11, e003837 (2023).
19.	Lakkis, J. et al. A multi-use deep learning method for CITE-seq and single-cell RNA-seq data integration with cell surface protein prediction and imputation. Nat. Mach. Intell. 4, 940–952 (2022).
20.	Yao, L. et al. Comprehensive Characterization of the Multiple Myeloma Immune Microenvironment Using Integrated scRNA-seq, CyTOF, and CITE-seq Analysis. Cancer Res. Commun. 2, 1255–1265 (2022).
21.	Ebili, H. O., Agboola, A. O. & Rakha, E. MSI-WES: a simple approach for microsatellite instability testing using whole exome sequencing. Future Oncol. Lond. Engl. 17, 3595–3606 (2021).
22.	Hao, S. et al. Synthetical lethality of Werner helicase and mismatch repair deficiency is mediated by p53 and PUMA in colon cancer. Proc. Natl. Acad. Sci. 119, e2211775119 (2022).
23.	Lin, K. X. et al. PD-1 and PD-L1 inhibitors in cold colorectal cancer: challenges and strategies. Cancer Immunol. Immunother. CII 72, 3875–3893 (2023).
24.	Wu, S. Z. et al. A single-cell and spatially resolved atlas of human breast cancers. Nat. Genet. 53, 1334–1347 (2021).
25.	Bonneville, R. et al. Landscape of Microsatellite Instability Across 39 Cancer Types. JCO Precis. Oncol. 1, PO.17.00073 (2017).
26.	Ratovomanana, T. et al. Performance of Next-Generation Sequencing for the Detection of Microsatellite Instability in Colorectal Cancer With Deficient DNA Mismatch Repair. Gastroenterology 161, 814-826.e7 (2021).
27.	Jin, X., Yan, J., Chen, C., Chen, Y. & Huang, W.-K. Integrated Analysis of Copy Number Variation, Microsatellite Instability, and Tumor Mutation Burden Identifies an 11-Gene Signature Predicting Survival in Breast Cancer. Front. Cell Dev. Biol. 9, 721505 (2021).
28.	Picco, G. et al. Novel WRN Helicase Inhibitors Selectively Target Microsatellite-Unstable Cancer Cells. Cancer Discov. 14, 1457–1475 (2024).
29.	Stoeckius, M. et al. Simultaneous epitope and transcriptome measurement in single cells. Nat. Methods 14, 865–868 (2017).
30.	Lazarchuk, P. et al. Werner syndrome RECQ helicase participates in and directs maintenance of the protein complexes of constitutive heterochromatin in proliferating human cells. 2024.01.29.577850 Preprint at https://doi.org/10.1101/2024.01.29.577850 (2024).
31.	Lieb, S. et al. Werner syndrome helicase is a selective vulnerability of microsatellite instability-high tumor cells. eLife 8, e43333.
32.	Tian, Y. et al. WRN loss accelerates abnormal adipocyte metabolism in Werner syndrome. Cell Biosci. 14, 1–19 (2024).
33.	Wu, J. et al. Lipidomic Risk Score to Enhance Cardiovascular Risk Stratification for Primary Prevention. JACC 84, 434–446 (2024).
34.	• Extract lipids using the Bligh and Dyer method with internal standards (e.g., d5-ceramide). - Google Search. https://www.google.com/search?client=firefox-b-1-d&q=%E2%80%A2+Extract+lipids+using+the+Bligh+and+Dyer+method+with+internal+standards+%28e.g.%2C+d5-ceramide%29.
35.	Peake, D. A., Kiyonami, R., Yokoi, Y. & Fukamachi, Y. Processing of a Complex Lipid Dataset for the NIST Inter-laborato and Plasma.
36.	Nie, H. et al. Targeting branched N-glycans and fucosylation sensitizes ovarian tumors to immune checkpoint blockade. Nat. Commun. 15, 2853 (2024).
37.	Lin, M.-C. et al. Targeting tumor O-glycosylation modulates cancer-immune-cell crosstalk and enhances anti-PD-1 immunotherapy in head and neck cancer. Mol. Oncol. 18, 350–368 (2024).
38.	Ren, X., Lin, S., Guan, F. & Kang, H. Glycosylation Targeting: A Paradigm Shift in Cancer Immunotherapy. Int. J. Biol. Sci. 20, 2607–2621 (2024).
39.	Prelich, G. Gene overexpression: uses, mechanisms, and interpretation. Genetics 190, 841–854 (2012).
40.	Jayaraman, J. et al. CAR-T design: Elements and their synergistic function. EBioMedicine 58, 102931 (2020).
41.	Guzman, G., Reed, M. R., Bielamowicz, K., Koss, B. & Rodriguez, A. CAR-T Therapies in Solid Tumors: Opportunities and Challenges. Curr. Oncol. Rep. 25, 479–489 (2023).
42.	Duncan, B. B., Dunbar, C. E. & Ishii, K. Applying a clinical lens to animal models of CAR-T cell therapies. Mol. Ther. Methods Clin. Dev. 27, 17–31 (2022).
43.	Lemoine, J., Ruella, M. & Houot, R. Born to survive: how cancer cells resist CAR T cell therapy. J. Hematol. Oncol.J Hematol Oncol 14, 1–12 (2021).
44.	Gust, J., Ponce, R., Liles, W. C., Garden, G. A. & Turtle, C. J. Cytokines in CAR T Cell–Associated Neurotoxicity. Front. Immunol. 11, 577027 (2020).
45.	Biery, D. N., Turicek, D. P., Diorio, C., Schroeder, B. A. & Shah, N. N. Need for standardization of cytokine profiling in CAR T cell therapy. Mol. Ther. J. Am. Soc. Gene Ther. 32, 2979–2983 (2024).
46.	Jia, P. et al. MSIsensor-RNA: Microsatellite Instability Detection for Bulk and Single-cell Gene Expression Data. Genomics Proteomics Bioinformatics 22, qzae004 (2024).
47.	Preprocessing and clustering. Scanpy https://scanpy.readthedocs.io/en/stable/tutorials/basics/clustering.html.
48.	Géron, A. Hands-on Machine Learning with Scikit-Learn, Keras, and TensorFlow: Concepts, Tools, and Techniques to Build Intelligent Systems. (O’Reilly, Beijing Boston Farnham Sebastopol Tokyo, 2023).
49.	He, Y., Vlaming, M., van Meerten, T. & Bremer, E. The Implementation of TNFRSF Co-Stimulatory Domains in CAR-T Cells for Optimal Functional Activity. Cancers 14, 299 (2022).
50.	Roselli, E. et al. 4-1BB and optimized CD28 co-stimulation enhances function of human mono-specific and bi-specific third-generation CAR T cells. J. Immunother. Cancer 9, e003354 (2021).
51.	Giordano, M. et al. The tumor necrosis factor alpha-induced protein 3 (TNFAIP3, A20) imposes a brake on antitumor activity of CD8 T cells. Proc. Natl. Acad. Sci. 111, 11115–11120 (2014).
52.	Rossetti, R. et al. Combination of genetically engineered T cells and immune checkpoint blockade for the treatment of cancer. Immunother. Adv. 2, ltac005 (2022).
53.	Sun, B. et al. Granzyme B-based CAR T cells block metastasis by eliminating circulating tumor cells. 2024.03.18.585442 Preprint at https://doi.org/10.1101/2024.03.18.585442 (2024).
54.	Duro-Sánchez, S., Alonso, M. R. & Arribas, J. Immunotherapies against HER2-Positive Breast Cancer. Cancers 15, 1069 (2023).
55.	Menyhart, O. et al. DUSP4 is associated with increased resistance against anti-HER2 therapy in breast cancer. Oncotarget 8, 77207–77218 (2017).
56.	Ros, J. et al. Immunotherapy for Colorectal Cancer with High Microsatellite Instability: The Ongoing Search for Biomarkers. Cancers 15, 4245 (2023).
57.	Adabi, E. et al. Enhanced conversion of T cells into CAR T cells by modulation of the MAPK/ERK pathway. Cell Rep. Med. 6, 101970 (2025).
58.	Daei Sorkhabi, A. et al. The current landscape of CAR T-cell therapy for solid tumors: Mechanisms, research progress, challenges, and counterstrategies. Front. Immunol. 14, 1113882 (2023).
59.	Kaczanowska, S. et al. Immune determinants of CAR-T cell expansion in solid tumor patients receiving GD2 CAR-T cell therapy. Cancer Cell 42, 35-51.e8 (2024).
60.	CAR T-Cell Therapy Market Size to Hit USD 127.53 Billion by 2033. BioSpace https://www.biospace.com/car-t-cell-therapy-market-size-to-hit-usd-127-53-billion-by-2033 (2024).